### PR TITLE
CSHARP-2913: Lift restriction on authSource without credentials

### PR DESCRIPTION
**Pull request by MikalaiMazurenka**
https://jira.mongodb.org/browse/CSHARP-2913
https://evergreen.mongodb.com/version/5ed06f363e8e864c9cd10a7c

Note: this PR also completes https://jira.mongodb.org/browse/CSHARP-3060

### CODE REVIEW

File: src/MongoDB.Driver/MongoUrl.cs [hunk 1]
```diff
@@ -272,7 +272,7 @@ public bool HasAuthenticationSettings
                     _username != null ||
                     _password != null ||
                     _authenticationMechanism != null ||
-                    _authenticationSource != null;
+                    _authenticationMechanismProperties.Count() > 0;
```
Comment (vincentkam) [inline src/MongoDB.Driver/MongoUrl.cs]: I understand why we're no longer checking `_authenticationSource !=null`.  Is the change for checking `_authenticationMechanismProperties.Count()` tied to spec change, or does it fix previously broken behavior?

Comment (MikalaiMazurenka) [reply src/MongoDB.Driver/MongoUrl.cs]: I think the specification justification is here: [Authentication source and URI database do not imply authentication](https://github.com/mongodb/specifications/blob/af6b8b14606d43f1938a2a8da23801f05c6cf4ab/source/auth/auth.rst#authentication-source-and-uri-database-do-not-imply-authentication)
> Similarly, the presence of the authSource option in the URI connection string without other credential data such as Userinfo or authentication parameters in connection options MUST NOT be interpreted as a request for authentication.

In my opinion, `authentication parameters` includes mechanism properties.
But essentially the line `_authenticationMechanismProperties.Count() > 0` doesn't fix any tests or any known issues. Neither it makes any tests fail.

**Review by vincentkam — APPROVED** [review]
LGTM!

File: src/MongoDB.Driver/MongoUrl.cs (hunk 1 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoUrl.cs]: I think the original idea of these lines could be: `_authenticationMechanismProperties` doesn't make sense without `_authenticationMechanism` itself, in other words, can we even have `_authenticationMechanismProperties` without `_authenticationMechanism`? If no so, I'm not sure that we need to add this change especially considering the fact that this change is not required by this ticket. Also, may this change be considered as a logic breaking change?

Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoUrl.cs]: Also, consider adding validation on null even though the current MongoUrlBuilder cannot return null, It's not a really good expectation to wait that a different class will never return `null` even if it's how it happens now

File: tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs [hunk 2]
```diff
@@ -75,9 +79,9 @@ private void AssertValid(MongoCredential mongoCredential, BsonDocument definitio
                 mongoCredential.Source.Should().Be(ValueToString(expectedCredential["source"]));
                 mongoCredential.Mechanism.Should().Be(ValueToString(expectedCredential["mechanism"]));
 
-                var authenticator = mongoCredential.ToAuthenticator();
-                if (authenticator is GssapiAuthenticator gssapiAuthenticator)
+                if (mongoCredential.Mechanism == GssapiAuthenticator.MechanismName)
                 {
+                    var gssapiAuthenticator = mongoCredential.ToAuthenticator() as GssapiAuthenticator;
                     expectedCredential.TryGetValue("mechanism_properties", out var expectedMechanismProperties);
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs]: this line is common for both code paths

File: tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs [hunk 3]
```diff
@@ -109,7 +113,19 @@ private void AssertValid(MongoCredential mongoCredential, BsonDocument definitio
                 }
                 else
                 {
-                    // Other authenticators do not contain mechanism properties
+                    if (expectedCredential.TryGetValue("mechanism_properties", out var expectedMechanismProperties))
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs]: I believe these two code paths can be merged into one. But I will rely on your and other reviewers' opinion here

**Review by DmitryLukyanov — COMMENTED** [review]
Some questions/minor suggestions

File: src/MongoDB.Driver/MongoUrl.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Driver/MongoUrl.cs]: Note that the following is *slightly* faster than `Count() > 0`:

    _authenticationMechanismProperties.Any();

Comment (rstam) [reply src/MongoDB.Driver/MongoUrl.cs]: I have the same question as Dmitry... do we even need to check `_authenticationModeProperties`?

Isn't it sufficient to just check `_authenticationMechanism  != null`?

File: tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs [hunk 4]
```diff
@@ -75,9 +79,9 @@ private void AssertValid(MongoCredential mongoCredential, BsonDocument definitio
                 mongoCredential.Source.Should().Be(ValueToString(expectedCredential["source"]));
                 mongoCredential.Mechanism.Should().Be(ValueToString(expectedCredential["mechanism"]));
 
-                var authenticator = mongoCredential.ToAuthenticator();
-                if (authenticator is GssapiAuthenticator gssapiAuthenticator)
+                if (mongoCredential.Mechanism == GssapiAuthenticator.MechanismName)
                 {
+                    var gssapiAuthenticator = mongoCredential.ToAuthenticator() as GssapiAuthenticator;
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs]: Use this instead:

    var gssapiAuthenticator = (GssapiAuthenticator)mongoCredential.ToAuthenticator();

File: tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs (hunk 3 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs]: There does seem to be some duplication.

But it's also slightly different. One path verifies the mechanism properties of the mechanism itself (but only for a GssapAuthenticator) while the other path verifies the mechanism properties of the credential itself.

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs (hunk 4 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs]: Done.

File: tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs (hunk 2 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs]: Done.

File: tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs (hunk 3 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs]: The difference is that for `GSSAPI` mechanism the code sets `SERVICE_NAME` and `CANONICALIZE_HOST_NAME` properties only after creating an `Authenticator` here: [GssapiAuthenticator.CreateMechanism()](https://github.com/mongodb/mongo-csharp-driver/blob/0d695d517fade7d70c207b1afbf10b5306a7d210/src/MongoDB.Driver.Core/Core/Authentication/GssapiAuthenticator.cs#L146). And these two properties are not taken directly from connection string but are expected by specification which is seen in the test: [should accept valid authSource (GSSAPI)](https://github.com/mongodb/specifications/blob/11970bc05b87ae58d8783d66b2d1eb29efdec04b/source/auth/tests/connection-string.json#L68).
Other mechanisms tests do not expect any properties besides those contained in connection string.

**Review by MikalaiMazurenka — COMMENTED** [review]
Ready for review!

File: src/MongoDB.Driver/MongoUrl.cs (hunk 1 above)
Comment (rstam) [reply src/MongoDB.Driver/MongoUrl.cs]: The following question has not been answered:

Isn't it sufficient to just check _authenticationMechanism != null?

**Review by rstam — CHANGES_REQUESTED** [review]
I unresolved a comment thread that contains a question by me that has not been answered yet.

**Review by DmitryLukyanov — CHANGES_REQUESTED** [review]
There is the unresolved comment

**Comment by MikalaiMazurenka** [comment]
I reached out to original spec author regarding `_authenticationMechanismProperties`. Hope it can shed some light onto this question.

File: src/MongoDB.Driver/MongoUrl.cs (hunk 1 above)
Comment (MikalaiMazurenka) [reply src/MongoDB.Driver/MongoUrl.cs]: I've checked with original specification author. He confirmed that he didn't intend any other meanings, so it's not necessary to check `_authenticationMechanismProperties`.
Updated the code to just skip `_authenticationSource` check.
Done.

**Review by MikalaiMazurenka — COMMENTED** [review]
Ready for review!

**Review by DmitryLukyanov — APPROVED** [review]
LGTM

**Review by vincentkam — APPROVED** [review]
Re-lgtming :)

**Review by rstam — APPROVED** [review]
LGTM

**Pull request closed without merging**